Add e2hs spec (#6)
eth-clients/e2store-format-specs: formats/e2hs.md

# E2HS files

 E2HS files are themselves e2store files. For more information on this format,
 see the [documentation](https:github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md).

 The overall structure of an E2HS file follows closely the structure of an Era file
 which contains consensus Layer data (and as a byproduct, EL data after the merge).

 The structure can be summarized through this definition:

> 	era1 := Version | block-tuple* | other-entries* | BlockIndex

Reviewer: ```suggestion
	e2hs := Version | block-tuple* | other-entries* | BlockIndex
```

> 	era1 := Version | block-tuple* | other-entries* | BlockIndex

Reviewer 2: What is "other-entries"? If I'm not mistaken, this isn't defined in any way.

Reviewer: ![image](https://github.com/user-attachments/assets/80b1654d-6bf3-46e0-a8db-2a21f01cc2b1)

This is in both era and era1

Reviewer 2: I wasn't aware of it :sweat: .
So I would add the same line in our spec as well (I wouldn't expect somebody reading about e2hs format to be familiar with other formats as well).

	block-tuple :=  CompressedHeaderWithProof | CompressedBody | CompressedReceipts

 Each basic element is its own entry:

	Version                     = { type: 0x6532, data: nil }
	CompressedHeaderWithProof   = { type: 0x0301, data: snappyFramed(ssz(header_with_proof)) }
	CompressedBody              = { type: 0x0400, data: snappyFramed(rlp(body)) }
	CompressedReceipts          = { type: 0x0500, data: snappyFramed(rlp(receipts)) }

> 	BlockIndex                  = { type: 0x3266, data: block-index }

Reviewer: ```suggestion
	BlockIndex                  = { type: 0x6632, data: block-index }
```

these bytes should be re-ordered

 The `header_with_proof` type definition can be found in the Portal Network [documentation](https://github.com/ethereum/portal-network-specs/blob/master/history/history-network.md#block-header).

 `BlockIndex` stores relative offsets to each compressed block entry. The
 format is:

	block-index := starting-number | index | index | index ... | count

 `starting-number` is the first block number in the archive. Every index is a
 defined relative to index's location in the file. The total number of block
 entries in the file is recorded in count.

>  The maximum number of blocks in an E2HS file is 8192.

Reviewer 2: Is 8192 the maximum number, or is that the exact number?

Regardless, I would put this information somewhere higher.

Reviewer: This is the exact number, era this would be the max number tho as slots can be skipped

Reviewer: also in era1 the last file isn't full, e2hs is unique compared to those 2 in that it would be exact

Author: yeah, agreed, every `e2hs` fille *MUST* have `8,192` contiguous el blocks. I'll update the spec. Though, in the previous closed pr @kdeme left this comment

> I think it is better to call 8192 blocks (or slots) an era, as epoch has another meaning on consensus side.

Any thoughts on this? We do use "epoch" accumulators for pre-merge validation... and imo "era" is just as ambiguous w/ regards to what it exactly refers to throughout all these file specs (`era` files / `era1` files etc.)

Reviewer: I think it is fine to say `8192 blocks`, because you use `era` you will need to define it in this spec for people who don't know what it means which would be redundent

Author: I guess we're also missing how to name these files. Any ideas? 

```
### File Naming Convention
`<config-name>-<era-number>.e2hs`

Where:
- `config-name`: Network configuration identifier (e.g., mainnet, sepolia)
- `era-number`: Sequential epoch identifier
```

Reviewer: Probably include a short hash of the file/data so `<config-name>-<era-number>-<short hash of e2hs file>.e2hs`

era1/era/e2ss all have some kind of hash in the name so might as well be consistent.

Reviewer 2: I don't know about era1 (it doesn't say in documentation), but era doesn't use the hash of a file itself. Instead, it uses the _"the first 4 bytes of the last historical root in the last state in the era file"_. This is useful to quickly check that it matches the canonical chain you are following.

In our case, I would rather use first 4 bytes of the first block_hash in the file (or something like that), then use hash of a file.

Author: Sounds good to me, seems like it'd be more useful, since the headers have proofs with them and I'm not sure we really need the validation that comes with a hash of the whole file

Reviewer: > In our case, I would rather use first 4 bytes of the first block_hash in the file (or something like that), then use hash of a file.

I think it would make more sense to use the last block header's hash in the file, as you can trace back to the first header that way, but you can't verify the last header is right from the first one

# E2HS files

 E2HS files are themselves e2store files. For more information on this format,
 see the [documentation](https:github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md).

 The overall structure of an E2HS file follows closely the structure of an Era file
 which contains consensus Layer data (and as a byproduct, EL data after the merge).

 The structure can be summarized through this definition:

	e2hs := Version | block-tuple* | other-entries* | BlockIndex
	block-tuple :=  CompressedHeaderWithProof | CompressedBody | CompressedReceipts

 Each basic element is its own entry:

> 	Version                     = { type: 0x6532, data: nil }

Reviewer 3: the type is a byte array ie better stick to the upstream array notation [`[0x65, 0x32]`](https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md#version) to avoid potential byte order issues.

	CompressedHeaderWithProof   = { type: 0x0301, data: snappyFramed(ssz(header_with_proof)) }
	CompressedBody              = { type: 0x0400, data: snappyFramed(rlp(body)) }
	CompressedReceipts          = { type: 0x0500, data: snappyFramed(rlp(receipts)) }
	BlockIndex                  = { type: 0x6632, data: block-index }

>  The `header_with_proof` type definition can be found in the Portal Network [documentation](https://github.com/ethereum/portal-network-specs/blob/master/history/history-network.md#block-header).

Reviewer 3: better link to a concrete commit throughout _or_ inline the types as branch links are subject to both changes and bitrot

 `BlockIndex` stores relative offsets to each compressed block entry. The
 format is:

	block-index := starting-number | index | index | index ... | count

 `starting-number` is the first block number in the archive. Every index is a
 defined relative to index's location in the file. The total number of block
 entries in the file is recorded in count.

 Every E2HS file must contain a contiguous section of 8192 blocks.

 ### File Naming Convention
 `<config-name>-<era-number>-<short-hash>.e2hs`

 Where:
 - `config-name`: Network configuration identifier (e.g., mainnet, sepolia) 
 - `era-number`: Sequential epoch identifier

>  - `short-hash`: first 4 bytes of the first block header included in this file

Reviewer: ```suggestion
 - `short-hash`: first 4 bytes of the last block header's hash included in this file
```

Shouldn't we do the last header in the file, as if you know the last header you can trace back, where the first header you can't trace forward?

Reviewer 3: either that, or the beacon block root that's used for verification

Author: went with the last header hash, since beacon block roots aren't possible for pre-merge `e2hs` files
